// expo-cli/build/credentials/api.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.IosApi = void 0;

function _xdl() {
  const data = require("@expo/xdl");

  _xdl = function () {
    return data;
  };

  return data;
}

function _findIndex() {
  const data = _interopRequireDefault(require("lodash/findIndex"));

  _findIndex = function () {
    return data;
  };

  return data;
}

function _find() {
  const data = _interopRequireDefault(require("lodash/find"));

  _find = function () {
    return data;
  };

  return data;
}

function _omit() {
  const data = _interopRequireDefault(require("lodash/omit"));

  _omit = function () {
    return data;
  };

  return data;
}

function _assign() {
  const data = _interopRequireDefault(require("lodash/assign"));

  _assign = function () {
    return data;
  };

  return data;
}

function _get() {
  const data = _interopRequireDefault(require("lodash/get"));

  _get = function () {
    return data;
  };

  return data;
}

function _log() {
  const data = _interopRequireDefault(require("../log"));

  _log = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class IosApi {
  constructor(user) {
    _defineProperty(this, "shouldRefetch", true);

    this.api = _xdl().ApiV2.clientForUser(user);
    this.credentials = {
      appCredentials: [],
      userCredentials: []
    };
  }

  async getAllCredentials() {
    if (this.shouldRefetch) {
      await this._fetchAllCredentials();
    }

    return this.credentials;
  }

  async _fetchAllCredentials() {
    (0, _log().default)('Fetching available credentials');
    this.credentials = await this.api.getAsync('credentials/ios');
    this.shouldRefetch = false;
  }

  async getDistCert(experienceName, bundleIdentifier) {
    if (this.shouldRefetch) {
      await this._fetchAllCredentials();
    }

    this._ensureAppCredentials(experienceName, bundleIdentifier);

    const credIndex = (0, _findIndex().default)(this.credentials.appCredentials, app => app.experienceName === experienceName && app.bundleIdentifier === bundleIdentifier);
    const distCertId = this.credentials.appCredentials[credIndex].distCredentialsId;

    if (!distCertId) {
      return null;
    }

    const distCert = this.credentials.userCredentials.find(cred => cred.id === distCertId);
    return distCert || null;
  }

  async createDistCert(credentials) {
    const {
      id
    } = await this.api.postAsync('credentials/ios/dist', {
      credentials
    });
    const newDistCert = { ...credentials,
      id,
      type: 'dist-cert'
    };
    this.credentials.userCredentials.push(newDistCert);
    return newDistCert;
  }

  async updateDistCert(credentialsId, credentials) {
    const {
      id
    } = await this.api.putAsync(`credentials/ios/dist/${credentialsId}`, {
      credentials
    });
    const updatedDistCert = { ...credentials,
      id,
      type: 'dist-cert'
    };
    const credIndex = (0, _findIndex().default)(this.credentials.userCredentials, ({
      id
    }) => id === credentialsId);
    this.credentials.userCredentials[credIndex] = updatedDistCert;
    return updatedDistCert;
  }

  async deleteDistCert(credentialsId) {
    await this.api.deleteAsync(`credentials/ios/dist/${credentialsId}`);
    this.credentials.userCredentials = this.credentials.userCredentials.filter(({
      id
    }) => id !== credentialsId);
    this.credentials.appCredentials = this.credentials.appCredentials.map(record => {
      if (record.distCredentialsId === credentialsId) {
        return (0, _omit().default)(record, 'distCredentialsId');
      }

      return record;
    });
  }

  async useDistCert(experienceName, bundleIdentifier, userCredentialsId) {
    await this.api.postAsync('credentials/ios/use/dist', {
      experienceName,
      bundleIdentifier,
      userCredentialsId
    });

    this._ensureAppCredentials(experienceName, bundleIdentifier);

    const credIndex = (0, _findIndex().default)(this.credentials.appCredentials, app => app.experienceName === experienceName && app.bundleIdentifier === bundleIdentifier);
    this.credentials.appCredentials[credIndex].distCredentialsId = userCredentialsId;
  }

  async createPushKey(credentials) {
    const {
      id
    } = await this.api.postAsync('credentials/ios/push', {
      credentials
    });
    const newPushKey = { ...credentials,
      id,
      type: 'push-key'
    };
    this.credentials.userCredentials.push(newPushKey);
    return newPushKey;
  }

  async updatePushKey(credentialsId, credentials) {
    const {
      id
    } = await this.api.putAsync(`credentials/ios/push/${credentialsId}`, {
      credentials
    });
    const updatedPushKey = { ...credentials,
      id,
      type: 'push-key'
    };
    const credIndex = (0, _findIndex().default)(this.credentials.userCredentials, ({
      id
    }) => id === credentialsId);
    this.credentials.userCredentials[credIndex] = updatedPushKey;
    return updatedPushKey;
  }

  async deletePushKey(credentialsId) {
    await this.api.deleteAsync(`credentials/ios/push/${credentialsId}`);
    this.credentials.userCredentials = this.credentials.userCredentials.filter(({
      id
    }) => id !== credentialsId);
    this.credentials.appCredentials = this.credentials.appCredentials.map(record => {
      if (record.pushCredentialsId === credentialsId) {
        return (0, _omit().default)(record, 'pushCredentialsId');
      }

      return record;
    });
  }

  async getPushKey(experienceName, bundleIdentifier) {
    if (this.shouldRefetch) {
      await this._fetchAllCredentials();
    }

    this._ensureAppCredentials(experienceName, bundleIdentifier);

    const credIndex = (0, _findIndex().default)(this.credentials.appCredentials, app => app.experienceName === experienceName && app.bundleIdentifier === bundleIdentifier);
    const pushKeyId = this.credentials.appCredentials[credIndex].pushCredentialsId;

    if (!pushKeyId) {
      return null;
    }

    const pushKey = this.credentials.userCredentials.find(cred => cred.id === pushKeyId);
    return pushKey || null;
  }

  async usePushKey(experienceName, bundleIdentifier, userCredentialsId) {
    await this.api.postAsync('credentials/ios/use/push', {
      experienceName,
      bundleIdentifier,
      userCredentialsId
    });

    this._ensureAppCredentials(experienceName, bundleIdentifier);

    const credIndex = (0, _findIndex().default)(this.credentials.appCredentials, app => app.experienceName === experienceName && app.bundleIdentifier === bundleIdentifier);
    this.credentials.appCredentials[credIndex].pushCredentialsId = userCredentialsId;
  }

  async deletePushCert(experienceName, bundleIdentifier) {
    await this.api.postAsync(`credentials/ios/pushCert/delete`, {
      experienceName,
      bundleIdentifier
    });
    const credIndex = (0, _findIndex().default)(this.credentials.appCredentials, app => app.experienceName === experienceName && app.bundleIdentifier === bundleIdentifier);
    this.credentials.appCredentials[credIndex].credentials = (0, _omit().default)(this.credentials.appCredentials[credIndex].credentials, ['pushId', 'pushP12', 'pushPassword']);
  }

  async updateProvisioningProfile(experienceName, bundleIdentifier, provisioningProfile, appleTeam) {
    await this.api.postAsync(`credentials/ios/provisioningProfile/update`, {
      experienceName,
      bundleIdentifier,
      credentials: { ...provisioningProfile,
        teamId: appleTeam.id,
        teamName: appleTeam.name
      }
    });
    const credIndex = (0, _findIndex().default)(this.credentials.appCredentials, app => app.experienceName === experienceName && app.bundleIdentifier === bundleIdentifier);
    (0, _assign().default)(this.credentials.appCredentials[credIndex].credentials, provisioningProfile);
    return provisioningProfile;
  }

  async getAppCredentials(experienceName, bundleIdentifier) {
    if (this.shouldRefetch) {
      await this._fetchAllCredentials();
    }

    this._ensureAppCredentials(experienceName, bundleIdentifier);

    return (0, _find().default)(this.credentials.appCredentials, app => app.experienceName === experienceName && app.bundleIdentifier === bundleIdentifier);
  }

  async getProvisioningProfile(experienceName, bundleIdentifier) {
    const appCredentials = await this.getAppCredentials(experienceName, bundleIdentifier);
    const provisioningProfileId = (0, _get().default)(appCredentials, 'credentials.provisioningProfileId');
    const provisioningProfile = (0, _get().default)(appCredentials, 'credentials.provisioningProfile');

    if (!provisioningProfileId || !provisioningProfile) {
      return null;
    }

    return {
      provisioningProfileId,
      provisioningProfile
    };
  }

  async deleteProvisioningProfile(experienceName, bundleIdentifier) {
    await this.api.postAsync(`credentials/ios/provisioningProfile/delete`, {
      experienceName,
      bundleIdentifier
    });
    const credIndex = (0, _findIndex().default)(this.credentials.appCredentials, app => app.experienceName === experienceName && app.bundleIdentifier === bundleIdentifier);
    this.credentials.appCredentials[credIndex].credentials = (0, _omit().default)(this.credentials.appCredentials[credIndex].credentials, ['provisioningProfile', 'provisioningProfileId']);
  }

  _ensureAppCredentials(experienceName, bundleIdentifier) {
    const exists = this.credentials.appCredentials.filter(i => i.experienceName === experienceName && i.bundleIdentifier === bundleIdentifier).length !== 0;

    if (!exists) {
      this.credentials.appCredentials.push({
        experienceName,
        bundleIdentifier,
        credentials: {}
      });
    }
  }

}

exports.IosApi = IosApi;
